README.md
** Weblogic Docker Container **

Build Docker Image for following versions
```
1 Weblogic 10.3.6 (coming soon)
2 Weblogic 12.2.1
3 Apache 2.4.18
```


CHANGELOG
Add change details here

weblogic.12.2.1/createDomain.sh
#!/bin/bash

########################################################################################
# Author: Habib Rangoonwala
# Created: 21-Jan-2016
# Updated: 29-Jan-2016
# Usage:
#  createDomain.sh <instancename>
#    <instancename> is used to as prefix/suffix for naming different components
########################################################################################


# to debug python, add -Dpython.verbose=debug
CONFIG_JVM_ARGS="${CONFIG_JVM_ARGS} -Dweblogic.security.SSL.ignoreHostnameVerification=true"

instancename=`echo $1|awk '{ print toupper($1) }'`
export NEW_DOMAIN_NAME="${instancename}_DOMAIN"
export CLUSTER_NAME="${instancename}_CLUSTER"
export ADMIN_HOST="wlsadmin$instancename"

#create the domain
${PAAS_HOME}/wlserver/common/bin/wlst.sh -skipWLSModuleScanning ${PAAS_HOME}/create-wls-domain.py

export DOMAIN_NAME=$NEW_DOMAIN_NAME

# unset $1 so that setDomainEnv.sh does not get any parameters
# due to some bug, if shift is not used setDomainEnv.sh will use $1 to launch managed server and fail
shift
. ${PAAS_HOME}/user_projects/domains/${DOMAIN_NAME}/bin/setDomainEnv.sh

#create boot.properties
mkdir -p ${PAAS_HOME}/user_projects/domains/${DOMAIN_NAME}/servers/AdminServer/security
echo "username=weblogic" > ${PAAS_HOME}/user_projects/domains/${DOMAIN_NAME}/servers/AdminServer/security/boot.properties 
echo "password=$ADMIN_PASSWORD" >> ${PAAS_HOME}/user_projects/domains/${DOMAIN_NAME}/servers/AdminServer/security/boot.properties 

#check how to persistently store this information
echo ". ${PAAS_HOME}/user_projects/domains/${DOMAIN_NAME}/bin/setDomainEnv.sh" >> ${PAAS_HOME}/.bashrc && \
echo "export PATH=$PATH:${PAAS_HOME}/wlserver/common/bin:${PAAS_HOME}/user_projects/domains/${DOMAIN_NAME}/bin" >> ${PAAS_HOME}/.bashrc

# pack the domain and change the permission so that other managed server process can read this domain template file
pack.sh -domain $DOMAIN_HOME -template $DOMAIN_HOME/${instancename}_DOMAIN.jar -template_name ${instancename}_DOMAIN -managed=true
chmod 775  $DOMAIN_HOME/${instancename}_DOMAIN.jar

startWebLogic.sh


weblogic.12.2.1/createServer.sh
#!/bin/bash

########################################################################################
# Author: Habib Rangoonwala
# Created: 21-Jan-2016
# Updated: 29-Jan-2016
# Usage:
#  createServer.sh <instancename>
#    <instancename> is used to as prefix/suffix for naming different components
########################################################################################


instancename="$1"
export DOMAIN_HOME="$PAAS_HOME/user_projects/domains/${instancename}_DOMAIN"

# start by unpacking the domain template created in createDomain.sh script
${PAAS_HOME}/wlserver/common/bin/unpack.sh -overwrite_domain true -domain $DOMAIN_HOME -template $DOMAIN_HOME/${instancename}_DOMAIN.jar

# now we have unpacked the domain so we can source domain environment file
. ${PAAS_HOME}/wls.env

export ADMIN_HOST="wlsadmin$instancename"
export CLUSTER_NAME="${instancename}_CLUSTER"

CONFIG_JVM_ARGS="${CONFIG_JVM_ARGS} -Dweblogic.security.SSL.ignoreHostnameVerification=true"
WLST="wlst.sh -skipWLSModuleScanning"

# Start Node Manager
nohup startNodeManager.sh > log.nm &
sleep 5

# Add a Machine to the AdminServer
$WLST ${PAAS_HOME}/add-wls-machine.py > `hostname`.log

# Wait and add a new Managed Server
$WLST ${PAAS_HOME}/add-managed-server.py >> `hostname`.log

# print log
tail -f `hostname`.log


weblogic.12.2.1/add-wls-machine.py
###################################
# Author: Habib Rangoonwala
# Created: 21-Jan-2016
# Updated: 28-Jan-2016
###################################

import os
import socket

# Functions
def editMode():
  edit()
  startEdit(3000000,300000,'true')

def editActivate():
  save()
  activate(block="true")

# Variables
# =========

# AdminServer details
username  = os.environ.get('ADMIN_USERNAME', 'weblogic')
password  = os.environ.get('ADMIN_PASSWORD')
adminhost = os.environ.get('ADMIN_HOST', 'wlsadmin')
adminport = os.environ.get('ADMIN_PORT', '8001')
domainhome = os.environ.get('DOMAIN_HOME', '')
domainname = os.environ.get('DOMAIN_NAME', 'DEFAULT_DOMAIN')

# NodeManager details
nmname = os.environ.get('NM_NAME', socket.gethostname())
nmhost = os.environ.get('NM_HOST', socket.gethostbyname(socket.gethostname()))
nmport = os.environ.get('NM_PORT', '5556')

#start Admin Server
#nmStart('AdminServer')
#nmServerStatus('AdminServer')

#startServer('AdminServer',domainname,'t3://' + adminhost + ':' + adminport,username,password,domainhome,'true',60000,'false')


# Connect to the AdminServer
# ==========================
connect(username, password, 't3://' + adminhost + ':' + adminport)

# Create a Machine
# ================
editMode()
cd('/')
cmo.createMachine(nmname)
cd('/Machines/' + nmname +'/NodeManager/' + nmname)
cmo.setListenPort(int(nmport))
cmo.setListenAddress(nmhost)
cmo.setNMType('Plain')
editActivate()

# Exit
# ====
exit()


apache.2.4.18/container-apachectl.sh
#!/bin/bash

#############################################
# Author: Habib Rangoonwala
# Created: 09-Jan-2016
# Updated: 28-Jan-2016
# Howto: 
# This script is set as CMD in Dockerfile
#############################################

# USE the trap if you need to also do manual cleanup after the service is stopped,
#     or need to start multiple services in the one container
trap "echo 'replace this with script to execute'" HUP INT QUIT KILL TERM

# start service in background here
echo "Adding this service to ETCD"

# find the docker_host_ip
DOCKER_HOST_IP=`/sbin/ip route|awk '/default/ { print $3 }'`
#curl -L -X PUT http://${DOCKER_HOST_IP}:2379/v2/keys/it/platform/apache -d value="$HOSTNAME"

# LD_LIBRARY_PATH  is required for mod_wl to work
# if not set you get this error
# Cannot load modules/mod_wl_24.so into server: libopmnsecure.so: cannot open shared object file: No such file or directory
export LD_LIBRARY_PATH=/usr/local/apache2/modules

echo "Starting Apache"
/usr/local/apache2/bin/apachectl start -D FOREGROUND

echo "[hit enter key to exit] or run 'docker stop <container>'"
read myval

# stop service and clean up here
echo "Stopping Apache"
/usr/local/apache2/bin/apachectl stop

echo "Removing this service from ETCD"
#curl -L -X DELETE http://${DOCKER_HOST_IP}:2379/v2/keys/it/platform/apache

echo "exited $0"


weblogic.12.2.1/add-managed-server.py
###################################
# Author: Habib Rangoonwala
# Created: 21-Jan-2016
# Updated: 29-Jan-2016
###################################

import os
import random
import string
import socket

# Functions
def randomName():
  return ''.join([random.choice(string.ascii_letters + string.digits) for n in xrange(6)])

def editMode():
  edit()
  startEdit(3000000,300000,'true')

def editActivate():
  save()
  activate(block="true")

# AdminServer details
username  = os.environ.get('ADMIN_USERNAME', 'weblogic')
password  = os.environ.get('ADMIN_PASSWORD')
adminhost = os.environ.get('ADMIN_HOST', 'wlsadmin')
adminport = os.environ.get('ADMIN_PORT', '8001')
cluster_name = os.environ.get("CLUSTER_NAME", "DEFAULT_CLUSTER")

# NodeManager details
nmname = os.environ.get('NM_NAME',socket.gethostname())

# ManagedServer details
msinternal = socket.gethostbyname(socket.gethostname())
msname = os.environ.get('MS_NAME', 'MS_' + socket.gethostname())
mshost = os.environ.get('MS_HOST', socket.gethostbyname(socket.gethostname()))
msport = os.environ.get('MS_PORT', '7001')
memargs = os.environ.get('USER_MEM_ARGS', '')

# Connect to the AdminServer
# ==========================
connect(username, password, 't3://' + adminhost + ':' + adminport)

# Create a ManagedServer
# ======================
editMode()
cd('/')
cmo.createServer(msname)

cd('/Servers/' + msname)
cmo.setMachine(getMBean('/Machines/' + nmname))
cmo.setCluster(getMBean('/Clusters/' + cluster_name))

# Default Channel for ManagedServer
# ---------------------------------
cmo.setListenAddress(msinternal)
cmo.setListenPort(int(msport))
cmo.setListenPortEnabled(true)
cmo.setExternalDNSName(mshost)

# Disable SSL for this ManagedServer
# ----------------------------------
cd('/Servers/' + msname + '/SSL/' + msname)
cmo.setEnabled(false)


# Custom Startup Parameters because NodeManager writes wrong AdminURL in startup.properties
# -----------------------------------------------------------------------------------------
cd('/Servers/' + msname + '/ServerStart/' + msname)
arguments = '-Djava.security.egd=file:/dev/./urandom -Dweblogic.Name=' + msname + ' -Dweblogic.management.server=http://' + adminhost + ':' + adminport + ' ' + memargs
cmo.setArguments(arguments)
editActivate()

# Start Managed Server
# ------------
try:
    start(msname, 'Server')
except:
    dumpStack()

# Exit
# =========
exit()


weblogic.12.2.1/create-wls-domain.py
###############################
# Author: Habib Rangoonwala
# Created: 21-Jan-2016
# Updated: 29-Jan-2016
###############################
domain_name  = "base_domain"
admin_port   = int(os.environ.get("ADMIN_PORT", "8001"))
admin_pass   = os.environ.get("ADMIN_PASSWORD", "welcome1")
cluster_name = os.environ.get("CLUSTER_NAME", "Cluster-Docker")
paas_home = os.environ.get("PAAS_HOME", "/paas")
domain_path  = paas_home + '/user_projects/domains/' + domain_name

# Open default domain template
# ======================
readTemplate(paas_home + "/wlserver/common/templates/wls/wls.jar")

# Configure the Administration Server and SSL port.
# =========================================================
cd('/Servers/AdminServer')
set('ListenAddress', '')
set('ListenPort', admin_port)

# Define the user password for weblogic
# =====================================
cd('/Security/' + domain_name + '/User/weblogic')
cmo.setPassword(admin_pass)

# Create a JMS Server
# ===================
cd('/')
create('myJMSServer', 'JMSServer')

# Create a JMS System resource
# ============================
cd('/')
create('myJmsSystemResource', 'JMSSystemResource')
cd('JMSSystemResource/myJmsSystemResource/JmsResource/NO_NAME_0')

# Create a JMS Queue and its subdeployment
# ========================================
myq = create('myQueue','Queue')
myq.setJNDIName('jms/myqueue')
myq.setSubDeploymentName('myQueueSubDeployment')

cd('/JMSSystemResource/myJmsSystemResource')
create('myQueueSubDeployment', 'SubDeployment')

# Target resources to the servers
# ===============================
cd('/')
assign('JMSServer', 'myJMSServer', 'Target', 'AdminServer')
assign('JMSSystemResource.SubDeployment', 'myJmsSystemResource.myQueueSubDeployment', 'Target', 'myJMSServer')

# Write the domain and close the domain template
# ==============================================
setOption('OverwriteDomain', 'true')
setOption('ServerStartMode','prod')

cd('/NMProperties')
set('ListenAddress','')
set('ListenPort',5556)
set('NativeVersionEnabled', 'false')
set('StartScriptEnabled', 'false')
set('SecureListener', 'false')

# Set the Node Manager user name and password
cd('/SecurityConfiguration/' + domain_name)
set('NodeManagerUsername', 'weblogic')
set('NodeManagerPasswordEncrypted', admin_pass)

# Define a WebLogic Cluster
# =========================
cd('/')
create(cluster_name, 'Cluster')

cd('/Clusters/' + cluster_name)
cmo.setClusterMessagingMode('unicast')

# Write Domain
# ============
setOption('OverwriteDomain', 'true')
domain_name  = os.environ.get("NEW_DOMAIN_NAME", "base_domain")
domain_path  = paas_home + '/user_projects/domains/' + domain_name
writeDomain(domain_path)
closeTemplate()

# Exit WLST
# =========
exit()
